### PR TITLE
Auto Stop Recording

### PR DESCRIPTION
**Pull request by vale981**
This little hack uses the beats menu in the context menu to set the maximum beats to record (Rounded up to bars). 

If no beat number is set, record works as usual. The beat number is reset upon track reset.
Cheers.

### CODE REVIEW

**Comment by vale981** [comment]
Well. It is not very intuitive, but UI savvy :).

I would add a new menu with bars not beats. 

On 25 March 2018 21:23:27 CEST, Harry van Haaren <notifications@github.com> wrote:
>Hi @vale981,
>
>This looks like a cool new feature! I like the idea, and I'd welcome
>the changes in general, but as you mention there is currently a "hacky"
>part in that we're re-using the "setBeats()" functionality...
>
>Do you think we should add a new item in the context menu for "Rec
>Beats" or similar? Or do you feel the current functionality is usable
>enough in its current form?
>
>Cheers, -H
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/openAVproductions/openAV-Luppp/pull/201#issuecomment-375996104

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by vale981** [comment]
P.S.: maybe we could implement one shot clips.

On 25 March 2018 21:23:27 CEST, Harry van Haaren <notifications@github.com> wrote:
>Hi @vale981,
>
>This looks like a cool new feature! I like the idea, and I'd welcome
>the changes in general, but as you mention there is currently a "hacky"
>part in that we're re-using the "setBeats()" functionality...
>
>Do you think we should add a new item in the context menu for "Rec
>Beats" or similar? Or do you feel the current functionality is usable
>enough in its current form?
>
>Cheers, -H
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/openAVproductions/openAV-Luppp/pull/201#issuecomment-375996104

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by harryhaaren** [comment]
OK cool, then wait for a few more commits that add the GUI elements to the context menu, and a new "EventType" message for LOOPER_REC_BEATS, which triggers the actual update to the looper clip class.

If you have questions or anything, just ask away here!

**Comment by harryhaaren** [comment]
Re One Shot clips - sure that sounds cool too - but lets finish one PR, and get that merged. Mixing multiple features in one PR isn't a good idea :)

**Comment by georgkrause** [comment]
I opened a new issue for one shot clips so you can concentrate here on auto record stop. #204

**Comment by vale981** [comment]
Shall i use callbacks for the right click menu, or that strcmp magic?

**Comment by vale981** [comment]
I am using callbacks.
BTW. All those events could be created very nicely with templates.

**Comment by vale981** [comment]
It ist just amazing that 
```
EventLooperBarsToRecord e(track->ID, track->getLastClipNum(), bars);
writeToDspRingbuffer( &e );
```

works. As far as i know `e` gets deallocated as soon as the function exits.

**Comment by georgkrause** [comment]
@vale981 when i click on custom and close the dialog without typing anything, luppp crashes. shell says: memory access error

**Comment by georgkrause** [comment]
Next time it would be much easier to use another branch than "master" for this, its kind of complicated to switch. or am maybe i am missing something.

**Comment by georgkrause** [comment]
fix works for me.

**Comment by vale981** [comment]
Neat

File: src/avtk/clipselector.cxx [hunk 1]
```diff
@@ -269,13 +299,25 @@ int ClipSelector::handle(int event)
...
+					{ "Bars to record",  0,   0, 0, FL_SUBMENU | FL_MENU_DIVIDER},
+					{"1 ", 0, setRecordBarsCb, (void*)1,0},
+					{"2 ", 0, setRecordBarsCb, (void*)2,0},
+					{"3 ", 0, setRecordBarsCb, (void*)3,0},
+					{"4 ", 0, setRecordBarsCb, (void*)4,0},
+					{"5 ", 0, setRecordBarsCb, (void*)5,0},
+					{"6 ", 0, setRecordBarsCb, (void*)6,0},
+					{"7 ", 0, setRecordBarsCb, (void*)7,0},
+					{"8 ", 0, setRecordBarsCb, (void*)8,0},
+					{"Custom", 0, setRecordBarsCb, (void*)-2,0},
```
Comment (harryhaaren) [inline src/avtk/clipselector.cxx]: I'm not sure how I feel about the (void*)-2, and -1. What purpose do Custom and Endless have?

**Comment by vale981** [comment]
I'd like to give the ClipState in the ui a pointer to the actual looperclip. So you can display how many beats a clip has. 

Is that ok to do and how would you do it. Call the logic directly from jack or through an event?

**Comment by vale981** [comment]
I prefer this methoder over handling the setbarstorecord event to change gui, because it is more transparent. 

But! It may not fit into the design structure.

**Comment by harryhaaren** [comment]
Hey @vale981, interesting approach... but not a good design choice in the Luppp context.

The core of Luppp is entirely split - the DSP code, and the UI code never access the same data. This allows (or *enforces* I suppose) the code to be lock-free, and real-time safe all the time.

What would happen if we de-allocate or re-allocate an instance? The pointer will be invalid - possibly for a short amount of time, but also possibly could cause a segfault. I'm aware that such a strong split between UI and DSP might seem overkill, but it does scale well. 

The message passing approach with the Event struct that Luppp uses throughout enables transfer of information between the two halves.

As such, I think the best approach is to add an Event struct type, and use that to signal the info back to the UI.

Hope this makes sense - any questions just let me know! -Harry

**Comment by vale981** [comment]
It does! I'll look into it.

**Comment by vale981** [comment]
I clearly see the advantage of this event driven approach for the means of realtime safety in the DSP code.
Certainly this code has some rough edges. But it is the first step in the right direction (so i hope).

Cheers.

**Comment by vale981** [comment]
Yes! I did change the indentation format on my main machine,  but forgot to do so on my laptop. It displays tabs and [4 Spaces] as exactly the same indent-level, so I didn't realize it.

Gonna fix that up.

**Comment by vale981** [comment]
We can run clang-format over it. But that would be another pull request.

**Comment by vale981** [comment]
This PR would fix #101.

**Comment by harryhaaren** [comment]
Re clang-format: yes great idea. I'd love to have a consistent method to keep the source clean. Do you have experience with setting up clang-format? A git commit hook to automatically fix it would be awesome :)

Yes noted that #101 will close on merging of this. Do you have any outstanding work todo on this PR, or it is ready for merge in your opinion?

Cheers, -H

**Comment by vale981** [comment]
I deem it ready.

**Comment by vale981** [comment]
But maybe sth. Comes to my mind.

**Comment by harryhaaren** [comment]
@vale981, I'll have a review now and if it looks good clicky on the merge button :)

[OT] Side note, I don't know if you use # IRC, but there's an #openav channel on irc.freenode.net where I hang out if i'm coding OpenAV stuff :)

**Comment by vale981** [comment]
So here we have it. Now it shows beats/bars.

**Comment by georgkrause** [comment]
I tested it. In general it works fine, its a great feature. Thank you for this!

- I think the display of the Beats and Bars is not that nice. Since the Clip shows only the first 8 characters of the name, there is some space on the right. Maybe you could use this?
lupppclip
- Maybe we should make the clip length display optional in the settings. But this could be another PR, just an idea right now. 
- I think it would be great to be able to change the default clip length. Currently it seems to be Endless and after the record is over, it gets resetted. Also in the settings we could define a default clip length for the current session which applies for all clips. 

Just some ideas from me, but i would be also happy it this could be merged as it is.

**Comment by vale981** [comment]
There is something wrong in ur picture

**Comment by vale981** [comment]
I suppose that I broke that

**Comment by georgkrause** [comment]
@vale981 will there be any further progress or do you think its ready? i could mark this as please merge, if you are done.

**Comment by vale981** [comment]
Sorry I was away for a bit.
I'll check in soon.

On 5 April 2018 10:29:42 CEST, Georg Krause <notifications@github.com> wrote:
>@vale981 will there be any further progress or do you think its ready?
>i could mark this as please merge, if you are done.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/openAVproductions/openAV-Luppp/pull/201#issuecomment-378859497

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by georgkrause** [comment]
dont feel pressure, just wanted to ask ;)

**Comment by vale981** [comment]
Default clip length would be inflexible. There is no good way of coping with this in the current single bar mode.

**Comment by vale981** [comment]
Having a Giada like sequencer would be a treat.

**Comment by georgkrause** [comment]
there might be an missunderstanding. i dont want to fix the length, just change the default length. currently its set to endless, which in fact isnt really length, but an setting for the record length. it would be nice to define a default, eg 8. you can manually stop the recording before anyway, right? So its like an maximum record length for the hole session.

**Comment by vale981** [comment]
Yes but one may want to record longer.... I would feel very uncomfortable if i could jam along indefinitely.

**Comment by vale981** [comment]
on smaller displays the text gets chopped off / overlaps with the title.

**Comment by georgkrause** [comment]
> Yes but one may want to record longer.... I would feel very uncomfortable if i could jam along indefinitely.

I dont want to prevent this, you could always change it to endless for each clip again. i just think it would be nice to be able to change the state which all clips have at the beginning.

> on smaller displays the text gets chopped off / overlaps with the title.

Currently if you make the window smaller the text jumps out of the clip. Two ideas:
- Put it on the right with a background, so the title gets cutted away if there are space problems
- Find another way to display, maybe more graphically (eg one dot for each bar, currently played bar in another color?)

Maybe we should move this display-thing to another commit since we need to think about again and maybe also add an setting to make it optionally. This way we could merge this one already.

**Comment by vale981** [comment]
Fixed it by trimming the text.

**Comment by vale981** [comment]
So. It is now color coded and the text trim is senistive to the beats/bars indicator.

**Comment by vale981** [comment]
So, I think it is good to merge :).

**Comment by georgkrause** [comment]
did a fix test, looks better but maybe we find a way to put the center of the numbers exactly over each other. and only the first track scene one is colored, others are white.

Sorry for complaining again and again, if you dont want to fix it i could do it in some days.

Thank you very much for the work, i really like the feature!

**Comment by vale981** [comment]
They are colored if there are 'beatsToRecord'

**Comment by vale981** [comment]
The numbers are aligned to the right. Center alignment is harder and doesn't do much good.

**Comment by vale981** [comment]
Got ya! Fixed that. There is another thingy that I am trying to get at though.

(Endless is always marked...)

**Comment by vale981** [comment]
Yes, Choosing another color would be better!

**Comment by harryhaaren** [comment]
@vale981 - please look into your editor white-space settings. There are mixed tabs/spaces and trailing whitespaces in the current code.

In general, most of the code itself is fine - I'll post a few review comments now.

File: src/avtk/clipselector.cxx [hunk 2]
```diff
@@ -90,6 +91,18 @@ void ClipSelector::clipName(int clip, std::string name)
 	redraw();
 }
 
+void ClipSelector::setClipBeats(int scene, int beats, bool isBeatsToRecord)
+{
+	clips[scene].setBeats(beats, isBeatsToRecord);
```
Comment (harryhaaren) [inline src/avtk/clipselector.cxx]: we need a `redraw()` call here to show the digit itself.

File: src/avtk/clipselector.cxx [hunk 3]
```diff
@@ -232,7 +277,49 @@ void ClipSelector::resize(int X, int Y, int W, int H)
...
+void setRecordBarsCb(Fl_Widget *w, void* data)
+{
+    if(!w || !data)
+        return;
+
+	ClipSelector *track = (ClipSelector*)w;
+	long bars = (long)data;
+	if(bars == -2){        
+		const char* answer = fl_input("Enter a custom number: ");
+        bars = atoi(answer);
```
Comment (harryhaaren) [inline src/avtk/clipselector.cxx]: error check *before* calling `atoi()` here.

File: src/avtk/clipselector.cxx [hunk 4]
```diff
@@ -232,7 +277,49 @@ void ClipSelector::resize(int X, int Y, int W, int H)
...
+        if(!answer || bars <= 0|| bars > MAX_BARS) {
```
Comment (harryhaaren) [inline src/avtk/clipselector.cxx]: remove `!answer` check from here, as it should be above

File: src/avtk/clipselector.cxx [hunk 5]
```diff
@@ -232,7 +277,49 @@ void ClipSelector::resize(int X, int Y, int W, int H)
...
+            bars = -1;
+            fl_message("Invalid Input. You may enter a number smaller than %i!", MAX_BARS);
```
Comment (harryhaaren) [inline src/avtk/clipselector.cxx]: reword message: "Please enter value between 1 and %`

File: src/avtk/clipselector.cxx [hunk 6]
```diff
@@ -232,7 +277,49 @@ void ClipSelector::resize(int X, int Y, int W, int H)
...
+        } else
+			bars = atoi(answer);
```
Comment (harryhaaren) [inline src/avtk/clipselector.cxx]: Else here isn't required, given the `atoi()` above will always already have executed

File: src/config.hxx [hunk 7]
```diff
@@ -65,6 +65,7 @@
 #define LUPPP_RETURN_WARNING  1
 #define LUPPP_RETURN_ERROR    2
 
+#define MAX_BARS 100
```
Comment (harryhaaren) [inline src/config.hxx]: Lets use 64 as MAX_BARS, 100 makes no sense in musical terms.

File: src/looper.cxx [hunk 8]
```diff
@@ -195,11 +199,11 @@ void Looper::pitchShift(int count, float* input, float* output)
 		int iTemp5 = int(fTemp4);
 		float out=output[0];
 		out += (float)(((1 - fTemp3) * (((fTemp4 - iTemp5) *
-		                                 fVec0[(IOTA-int((int((1 + iTemp5)) & 65535)))&65535]) + ((0 - ((
-		                                                 fRec0[0] + fSlow3) - iTemp5)) * fVec0[(IOTA-int((iTemp5 & 65535)))
-		                                                                 &65535]))) + (fTemp3 * (((fRec0[0] - iTemp1) * fVec0[(IOTA-int((int(
-		                                                                                 iTemp2) & 65535)))&65535]) + ((iTemp2 - fRec0[0]) * fVec0[(IOTA-int((
-		                                                                                                 iTemp1 & 65535)))&65535]))));
+										 fVec0[(IOTA-int((int((1 + iTemp5)) & 65535)))&65535]) + ((0 - ((
```
Comment (harryhaaren) [inline src/looper.cxx]: Some whitespace script seems to have reformatted this - see if you can remove this change as it isn't related to the PR.

File: src/avtk/clipselector.cxx (hunk 3 above)
Comment (vale981) [reply src/avtk/clipselector.cxx]: sorry. that was a quick hack that i forgot

**Comment by vale981** [comment]
Do not merge. There is a bug.

**Comment by vale981** [comment]
Here we go! All good now.

**Comment by vale981** [comment]
Something actually undid the major feature ^^.

**Comment by georgkrause** [comment]
> Something actually undid the major feature ^^.

@vale981 what do you mean?

**Comment by coderkun** [comment]
I can confirm that the major feature does not work: The recording doesn’t stop for me but runs endlessly …

**Comment by georgkrause** [comment]
Its working here with the latest commit in this thread.

**Comment by vale981** [comment]
It is working

**Comment by vale981** [comment]
This is one of Ask the last days I am able to work on this as my semester is going to start very soon.

 If there are there are any wishes. Voice them now.

**Comment by vale981** [comment]
@coderkun I shouldn't have put this comment in after the commit.

**Comment by georgkrause** [comment]
If you changed everything Harry requested this should be fine. Please be patient ;)

If you want you could add comments for the doxygen documentation to your code

**Comment by coderkun** [comment]
I can confirm that the feature is working (again). Would be good to squash the commits though.

**Comment by vale981** [comment]
Sorry for all the whitespace issues. I was lazy and should have limited that script to the files I changed. Nevertheless I endorse clang-format as it is a good way to keep the style consistent, even with devs who ain't use no tabs anywhere in their system.

File: src/avtk/clipselector.cxx [hunk 9]
```diff
@@ -245,11 +337,7 @@ int ClipSelector::handle(int event)
 	case FL_PUSH:
 		highlight = 1;
 		{
-			// calculate the clicked clip number
-			int clipHeight = (h / numClips);
-			int clipNum = ( (Fl::event_y() ) - y ) / clipHeight;
-			if (clipNum >= numClips)
-				clipNum = numClips -1; // fix for clicking the lowest pixel
+			findClipNum();
```
[the remark this answers is not in the document]
Comment (vale981) [reply src/avtk/clipselector.cxx]: It does something as this functionality saves the clicked clip num. Would be clearer with an assignment though. Or at least a rename.

File: src/avtk/clipselector.cxx [hunk 10]
```diff
@@ -196,16 +210,48 @@ void ClipSelector::draw()
...
 			// clip name
 			cairo_move_to( cr, x + clipHeight + 5, drawY + textHeight );
 			cairo_set_source_rgba( cr, 255 / 255.f, 255 / 255.f , 255 / 255.f , 0.9 );
 			cairo_set_font_size( cr, 11 );
 
-			std::string tmp = clips[i].getName().substr(0,8);
+            std::string tmp = clips[i].getName();
+
+            // trim the names
+            while(getCairoTextWith(cr, tmp.c_str()) > clipWidth - (clipHeight + 15 + beatLen)){
```
[the remark this answers is not in the document]
Comment (vale981) [reply src/avtk/clipselector.cxx]: Setting those constants somewhere Central and wrapping it in a method would make it clearer. 

Maybe Cairo has a better method for trimming text.

A monospaced font would also be a solution.

File: src/avtk/clipselector.cxx (hunk 9 above)
[the remark this answers is not in the document]
Comment (vale981) [reply src/avtk/clipselector.cxx]: Measuring M as the widest character ....

**Comment by vale981** [comment]
Well I actually can't tell why I haven't looked at  the diffs.

File: src/avtk/clipselector.cxx (hunk 10 above)
[the remark this answers is not in the document]
Comment (vale981) [reply src/avtk/clipselector.cxx]: I have tested this with the widest character M but it doesnt give a nice result.

**Comment by vale981** [comment]
I'll close this for now.

**Comment by georgkrause** [comment]
Will you file another PR with this functionallity?

**Comment by vale981** [comment]
A squashed one probably

**Pull request closed without merging**